ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,278 @@
+{
+    NGRAPH_OP_SCOPE(v7_Einsum_extract_labels);
+
+    labels.clear();
+    auto subscript_length = subscript.length();
+    for (size_t ch_idx = 0; ch_idx < subscript_length; ++ch_idx)
+    {
+        if (std::isalpha(subscript[ch_idx]))
+        {
+            labels.push_back(subscript.substr(ch_idx, 1));
+        }
+        else if (((subscript_length - ch_idx) > 2) &&
+                 (subscript.substr(ch_idx, 3).compare("...") == 0))
+        {
+            labels.push_back("...");
+            // make additional increment since ellipsis consists of three dots.
+            ch_idx += 2;
+        }
+    }
+}
+
+void op::v7::Einsum::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v7_Einsum_validate_and_infer_types);
+
+    // check that Einsum operation has at least one input
+    auto num_inputs = get_input_size();
+    NODE_VALIDATION_CHECK(this, num_inputs > 0, "Einsum must have at least one input.");
+
+    // check that all inputs have the same type and the type is numeric
+    const auto& input_type_0 = get_input_element_type(0);
+    NODE_VALIDATION_CHECK(this,
+                          input_type_0.is_real() || input_type_0.is_integral_number(),
+                          "The input type for Einsum operation must be numeric.");
+    for (size_t input_idx = 1; input_idx < num_inputs; ++input_idx)
+    {
+        const auto& input_type_i = get_input_element_type(input_idx);
+        NODE_VALIDATION_CHECK(this,
+                              input_type_0 == input_type_i,
+                              "Inputs to Einsum operation must have the same type.");
+    }
+
+    // check that equation has correct format and extract input and output subscripts
+    std::vector<std::string> input_subscripts;
+    std::string output_subscript;
+    NODE_VALIDATION_CHECK(this,
+                          parse_equation(input_subscripts, output_subscript) == true,
+                          "Equation must be strictly in explicit or implicit mode.");
+    NODE_VALIDATION_CHECK(
+        this,
+        input_subscripts.size() == num_inputs,
+        "Equation must contain a number of subscripts equal to a number of Einsum inputs.");
+
+    // create a dictionary with dimension sizes (or ranges in case dynamic shapes) for each label
+    // and check their compatibility in case repeating labels
+    unordered_map<string, PartialShape> label_to_shape;
+    label_to_shape.clear();
+
+    for (size_t input_idx = 0; input_idx < num_inputs; ++input_idx)
+    {
+        const auto& pshape = get_input_partial_shape(input_idx);
+        std::vector<std::string> labels;
+        extract_labels(input_subscripts[input_idx], labels);
+
+        if (pshape.rank().is_static())
+        {
+            size_t input_rank = pshape.rank().get_length();
+            // check that a rank is equal to a number of labels
+            NODE_VALIDATION_CHECK(this,
+                                  input_rank >= labels.size(),
```

Comment:
Either the condition or the comments in the code must be updated. Looks like the comment must be updated because the ellipsis label could correspond to several input dimensions.

Reply:
Fixed.